incubator/binding-mqtt-kafka/src/main/java/io/aklivity/zilla/runtime/binding/mqtt/kafka/internal/stream/MqttKafkaSessionFactory.java

```diff
@@ -741,30 +991,132 @@ private void onKafkaData(
+                    }
+                    MqttWillSignalFW willSignal =
+                        mqttWillSignalRO.tryWrap(payload.buffer(), payload.offset(), payload.limit());
+
+                    long deliverAt = willSignal.deliverAt();
+                    final String16FW clientId = willSignal.clientId();
+
+                    if (deliverAt == MqttWillDeliverAt.UNKNOWN.value())
+                    {
+                        if (!instanceId.getInstanceId().equals(willSignal.instanceId()))
+                        {
+                            deliverAt = supplyTime.get() + SECONDS.toMillis(willSignal.delay());
+                        }
+                        else
+                        {
+                            break fetchWill;
+                        }
+                    }
+
+                    KafkaFetchWillProxy willFetcher = new KafkaFetchWillProxy(originId, routedId, this, sessionsTopic,
+                        clientId, willSignal.willId().asString(), willSignal.lifetimeId().asString(), deliverAt);
+                    willFetcher.doKafkaBegin(traceId, authorization, 0, willSignal.lifetimeId());
+                    willFetchers.put(clientId, willFetcher);
+                }
             }
         }
 
-        protected abstract void handleKafkaData(DataFW data);
+        private void onKafkaFlush(
+            FlushFW flush)
+        {
+            final long sequence = flush.sequence();
+            final long acknowledge = flush.acknowledge();
 
-        protected abstract void onKafkaWindow(WindowFW window);
+            assert acknowledge <= sequence;
+            assert sequence >= replySeq;
 
-        protected void onKafkaEnd(
-            EndFW end)
-        {
+            replySeq = sequence;
+
+            assert replyAck <= replySeq;
+
+            final OctetsFW extension = flush.extension();
+            final ExtensionFW flushEx = extension.get(extensionRO::tryWrap);
+            final KafkaFlushExFW kafkaFlushEx =
+                flushEx != null && flushEx.typeId() == kafkaTypeId ? extension.get(kafkaFlushExRO::tryWrap) : null;
+            final KafkaMergedFlushExFW kafkaMergedFlushEx =
+                kafkaFlushEx != null && kafkaFlushEx.kind() == KafkaDataExFW.KIND_MERGED ? kafkaFlushEx.merged() : null;
+            final Array32FW<KafkaOffsetFW> progress = kafkaMergedFlushEx != null ? kafkaMergedFlushEx.progress() : null;
+
+            if (progress != null)
+            {
+                final IntHashSet newPartitions = new IntHashSet();
+                progress.forEach(p -> newPartitions.add(p.partitionId()));
+                if (!newPartitions.equals(partitions))
+                {
+                    instanceId.regenerate();
+                    partitions = newPartitions;
+                }
+            }
+            else
+            {
+                //TODO: this means we lost all partitions during the assignment, so we need to regenerate, right?
```

Comment:
I think this means we received a `FLUSH` with different or missing `KafkaMergedFlushEx` extension, right?

Perhaps we can afford to be more strict here, and not try to tolerate an edge case we don't need,